# Working with data in Python day 3 - Exercises with answers

## Exercise 1

#### Question 1

##### Import `pandas` package as `pd` 
##### Import `numpy` as `np`
##### Create a pandas Series object from a list of groceries below. Assign it to `shopping_series` variable and print it.
- milk
- eggs
- peppers
- cheese
- tomatoes

##### Print out only the values of the Series.

#### Answer:

In [50]:
import pandas as pd
import numpy as np

In [51]:
# Create pandas series object
shopping_series = pd.Series(["milk", "eggs", "peppers", "cheese", "tomatoes"])
print(shopping_series)

0        milk
1        eggs
2     peppers
3      cheese
4    tomatoes
dtype: object


In [52]:
# Print out values of series
print(shopping_series.values)

['milk' 'eggs' 'peppers' 'cheese' 'tomatoes']


#### Question 2
##### Create a pandas Date Series object starting from `January 1st 2019` and ending with `January 1st 2020` separated by a month. 
##### Assign the Series to `monthly_series` variable. Print the result.     
<br>

##### Create another pandas Date Series object starting from `January 1st 2019` and covering 12 periods separated by a month. 
##### Assign the Series to `monthly_series_by_period` variable. Print the result.
##### Are the two Series equivalent?

#### Answer:

In [53]:
monthly_series = pd.date_range(start = '20190101', end = '20200101', freq = 'M')
print(monthly_series)

DatetimeIndex(['2019-01-31', '2019-02-28', '2019-03-31', '2019-04-30',
               '2019-05-31', '2019-06-30', '2019-07-31', '2019-08-31',
               '2019-09-30', '2019-10-31', '2019-11-30', '2019-12-31'],
              dtype='datetime64[ns]', freq='M')


In [54]:
monthly_series_by_period = pd.date_range(start = '20190101', freq = 'M', periods = 12)
print(monthly_series_by_period)

DatetimeIndex(['2019-01-31', '2019-02-28', '2019-03-31', '2019-04-30',
               '2019-05-31', '2019-06-30', '2019-07-31', '2019-08-31',
               '2019-09-30', '2019-10-31', '2019-11-30', '2019-12-31'],
              dtype='datetime64[ns]', freq='M')


Yes, the two series objects are equivalent!

#### Question 3
##### Create a pandas Date Series object starting from `January 1st 2019` and ending with `January 3rd 2019` separated by an hour. Assign the Series to `hourly_series` variable. 
##### Print the result.

#### Answer:

In [55]:
hourly_series = pd.date_range(start='20190101', end='20190103', freq='H')

In [56]:
print(hourly_series)

DatetimeIndex(['2019-01-01 00:00:00', '2019-01-01 01:00:00',
               '2019-01-01 02:00:00', '2019-01-01 03:00:00',
               '2019-01-01 04:00:00', '2019-01-01 05:00:00',
               '2019-01-01 06:00:00', '2019-01-01 07:00:00',
               '2019-01-01 08:00:00', '2019-01-01 09:00:00',
               '2019-01-01 10:00:00', '2019-01-01 11:00:00',
               '2019-01-01 12:00:00', '2019-01-01 13:00:00',
               '2019-01-01 14:00:00', '2019-01-01 15:00:00',
               '2019-01-01 16:00:00', '2019-01-01 17:00:00',
               '2019-01-01 18:00:00', '2019-01-01 19:00:00',
               '2019-01-01 20:00:00', '2019-01-01 21:00:00',
               '2019-01-01 22:00:00', '2019-01-01 23:00:00',
               '2019-01-02 00:00:00', '2019-01-02 01:00:00',
               '2019-01-02 02:00:00', '2019-01-02 03:00:00',
               '2019-01-02 04:00:00', '2019-01-02 05:00:00',
               '2019-01-02 06:00:00', '2019-01-02 07:00:00',
               '2019-01-

#### Question 4
##### Take `numpy` array `abs_values` created in yesterday's exercises as follows:

In [57]:
abs_values = abs(np.array([5, -8, -66, 4]))

#### Convert `abs_values` into a pandas Series object. 
##### Assign it to variable `abs_series`.
##### Print the result.
##### Then print the results of the following operations:
- Get the shape of the Series
- Find the mean of the Series
- Find the median of the Series
- Find the standard deviation of the Series
- Get unique values in the Series
- Get the number of unique values in the Series
- Get counts of unique values in the Series


#### Answer:

In [58]:
abs_series = pd.Series(abs_values)
print(abs_series)

0     5
1     8
2    66
3     4
dtype: int64


In [59]:
print(abs_series.shape)

(4,)


In [60]:
print(abs_series.mean())

20.75


In [61]:
print(abs_series.median())

6.5


In [62]:
print(abs_series.std())

30.21451086260816


In [63]:
print(abs_series.unique())

[ 5  8 66  4]


In [64]:
print(abs_series.nunique())

4


In [65]:
print(abs_series.value_counts())

5     1
4     1
66    1
8     1
dtype: int64


#### Question 5 

##### For the same series `abs_series`, print out the results of the following operations:

- Find the position of the minimum value in the Series
- Find the position of the maximum value in the Series
- Rank items in the Series in descending order
- Sort values in the Series
- Find sum of the Series
- Find cumulative sum of the Series


#### Answer:

In [66]:
print(abs_series.idxmin()) 

3


In [67]:
print(abs_series.idxmax())

2


In [68]:
print(abs_series.rank(ascending = False))

0    3.0
1    2.0
2    1.0
3    4.0
dtype: float64


In [69]:
print(abs_series.sort_values())

3     4
0     5
1     8
2    66
dtype: int64


In [70]:
print(abs_series.sum())

83


In [71]:
print(abs_series.cumsum())

0     5
1    13
2    79
3    83
dtype: int64


## Exercise 2

#### Question 1

##### Use the `shopping_series` series created in Exercise 1.
##### Create a Series object called `prices` that contains the following values: [2.5, 3.99, 1.95, 5.99, 1.29].
##### Create a Series object called `num_items` that contains the following values: [1, 1, 2, 1, 3].
##### Combine the three Series objects described above into a single DataFrame called `shopping_cart` naming the columns like so:
- `item_name`
- `item_price`
- `item_quantity`

##### Print the first few entries (i.e. the `head` of the DataFrame).

#### Answer:

In [72]:
prices = pd.Series([2.5, 3.99, 1.95, 5.99, 1.29])
num_items = pd.Series([1, 1, 2, 1, 3])

In [73]:
shopping_cart = pd.DataFrame({'item_name': shopping_series, 'item_price': prices, 'item_quantity': num_items})
print(shopping_cart.head())

  item_name  item_price  item_quantity
0      milk        2.50              1
1      eggs        3.99              1
2   peppers        1.95              2
3    cheese        5.99              1
4  tomatoes        1.29              3


#### Question 2
##### Check the following characteristics of the `shopping_cart` object:
- Type
- Shape
- Show columns of the DataFrame
- Show information of the DataFrame
- Describe the DataFrame

##### How many columns are described in the last output? Why do you think that's the case?
##### What is the 75th percentile of the column `item_price`?
##### What is the standard deviation of the column `item_quantity`?

#### Answer:

In [74]:
print(type(shopping_cart))

<class 'pandas.core.frame.DataFrame'>


In [75]:
print(shopping_cart.shape)

(5, 3)


In [77]:
print(shopping_cart.columns)

Index(['item_name', 'item_price', 'item_quantity'], dtype='object')


In [78]:
print(shopping_cart.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   item_name      5 non-null      object 
 1   item_price     5 non-null      float64
 2   item_quantity  5 non-null      int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 248.0+ bytes
None


In [79]:
print(shopping_cart.describe())

       item_price  item_quantity
count     5.00000       5.000000
mean      3.14400       1.600000
std       1.87704       0.894427
min       1.29000       1.000000
25%       1.95000       1.000000
50%       2.50000       1.000000
75%       3.99000       2.000000
max       5.99000       3.000000


Only `2` columns are described in the `describe()` output since these are the only 2 numerical columns.

The 75th percentile of the column `item_price` is `3.99`

The standard deviation of the column `item_quantity` is 0.894427

#### Question 3
##### Perform the following on `shopping_cart` DataFrame:
- Extract column `item_price`
- Extract columns `item_name` and`item_quantity`
- Extract the 3rd row


#### Answer:

In [80]:
print(shopping_cart['item_price'])

0    2.50
1    3.99
2    1.95
3    5.99
4    1.29
Name: item_price, dtype: float64


In [81]:
print(shopping_cart[['item_name', 'item_quantity']])

  item_name  item_quantity
0      milk              1
1      eggs              1
2   peppers              2
3    cheese              1
4  tomatoes              3


In [82]:
print(shopping_cart.iloc[2])

item_name        peppers
item_price          1.95
item_quantity          2
Name: 2, dtype: object


#### Question 4
##### Set index of `shopping_cart` DataFrame to `item_name`. Print the updated DataFrame.
##### Print only the index of the `shopping_cart` DataFrame.
##### Look up the row of the `shopping_cart` by using index value 'eggs'.
##### Reset index of the `shopping_cart`. Print the updated DataFrame.

#### Answer:

In [83]:
shopping_cart = shopping_cart.set_index('item_name')
print(shopping_cart)

           item_price  item_quantity
item_name                           
milk             2.50              1
eggs             3.99              1
peppers          1.95              2
cheese           5.99              1
tomatoes         1.29              3


In [84]:
print(shopping_cart.index)

Index(['milk', 'eggs', 'peppers', 'cheese', 'tomatoes'], dtype='object', name='item_name')


In [85]:
print(shopping_cart.loc['eggs'])

item_price       3.99
item_quantity    1.00
Name: eggs, dtype: float64


In [86]:
shopping_cart = shopping_cart.reset_index()
print(shopping_cart)

  item_name  item_price  item_quantity
0      milk        2.50              1
1      eggs        3.99              1
2   peppers        1.95              2
3    cheese        5.99              1
4  tomatoes        1.29              3


#### Question 5

##### Import the package `os`
##### Set the working directory to data directory.
##### Print the working directory.

#### Answer:

In [87]:
import os

In [ ]:
# Set `main_dir` to the location of your `skill-soft` folder (for Linux).
main_dir = "/home/[username]/Desktop/skill-soft"
# Windows only.
main_dir = "C:/Users/[user_name]/Desktop/skill-soft"
# Mac only.
main_dir = "/Users/[username]/Desktop/skill-soft"

data_dir = main_dir + "/data/"

In [89]:
# Set working directory.
os.chdir(data_dir)

In [ ]:
# Check working directory.
print(os.getcwd())

#### Question 6
##### Read in 'TB_data.csv' into a variable called `tb_data` using pandas.
##### Print the first 20 rows of the DataFrame.

#### Answer:

In [91]:
tb_data = pd.read_csv('TB_data.csv')
print(tb_data.head(20))

        country iso2 iso3  iso_numeric  year measure unit age_group sex  \
0   Afghanistan   AF  AFG            4  2017     inc  num      0-14   a   
1   Afghanistan   AF  AFG            4  2017     inc  num      0-14   f   
2   Afghanistan   AF  AFG            4  2017     inc  num      0-14   m   
3   Afghanistan   AF  AFG            4  2017     inc  num       0-4   f   
4   Afghanistan   AF  AFG            4  2017     inc  num       0-4   m   
5   Afghanistan   AF  AFG            4  2017     inc  num     15-24   f   
6   Afghanistan   AF  AFG            4  2017     inc  num     15-24   m   
7   Afghanistan   AF  AFG            4  2017     inc  num    15plus   a   
8   Afghanistan   AF  AFG            4  2017     inc  num    15plus   f   
9   Afghanistan   AF  AFG            4  2017     inc  num    15plus   m   
10  Afghanistan   AF  AFG            4  2017     inc  num     25-34   f   
11  Afghanistan   AF  AFG            4  2017     inc  num     25-34   m   
12  Afghanistan   AF  AFG